aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp

```diff
@@ -563,6 +422,280 @@ void upsample_trilinear3d_backward_kernel_impl(
+               is_all_zero_stride<out_ndims, -1, scalar_t, index_t>(&strides[2]))) {
+      // contiguous channels-last case
+      basic_loop<scalar_t, index_t, out_ndims>(data, strides, n);
+    } else {
+      // fallback
+      basic_loop<scalar_t, index_t, out_ndims>(data, strides, n);
+    }
+  };
+  iter.for_each(loop);
+}
+
+template<typename index_t, typename scalar_t>
+std::vector<Tensor> ti_compute_indices_weights_linear(
+  int64_t input_size, int64_t output_size, int64_t stride, int64_t ndims, int64_t reshape_dim, 
+  bool align_corners, const c10::optional<double> opt_scale
+) {
+
+  scalar_t scale = area_pixel_compute_scale<scalar_t>(input_size, output_size, align_corners, opt_scale);
+
+  std::vector<Tensor> output;
+  auto new_shape = std::vector<int64_t>(ndims, 1);
+  new_shape[reshape_dim] = output_size;
+
+  output.emplace_back(empty(new_shape, CPU(c10::CppTypeToScalarType<index_t>())));
+  output.emplace_back(empty(new_shape, CPU(c10::CppTypeToScalarType<scalar_t>())));  
+  output.emplace_back(empty(new_shape, CPU(c10::CppTypeToScalarType<index_t>())));
+  output.emplace_back(empty(new_shape, CPU(c10::CppTypeToScalarType<scalar_t>())));
+
+  auto input_index0_ptr = output[0].data_ptr<index_t>();
+  auto lambda0_ptr = output[1].data_ptr<scalar_t>();
+  auto input_index1_ptr = output[2].data_ptr<index_t>();
+  auto lambda1_ptr = output[3].data_ptr<scalar_t>();
+
+  double xd;
+  int64_t xl;
+  
+  for (int64_t i=0; i<output_size; i++) {
+
+    compute_source_index_and_lambda<scalar_t, index_t>(
+      input_index0_ptr[i], input_index1_ptr[i],
+      lambda0_ptr[i], lambda1_ptr[i],
+      scale, i, input_size, output_size, align_corners
+    );
+    // put stride into indices
+    // index values correspond to input indices (0, 1, 2, 3, ...)
+    // when multiplied by input stride, maximum possible value
+    // input_size[dim-1] * input_size[dim-2] * ... for the given dimension.
+    input_index0_ptr[i] *= stride;
+    input_index1_ptr[i] *= stride;
+  }
+  return output;
+}
+
+
+// Upsampling linear interpolation kernel for N-d case.
+// Internally, it uses TensorIterator to optimize the computations.
+// Output is constructed inside the function and is a contiguous tensor.
+// - index_t is template type for input index: int32_t or int64_t
+// - out_ndims is the number of interpolated dims: 1, 2, 3
+// - scale_type is template type for scales, typically c10::optional<double>
+template <typename index_t, int out_ndims, typename scale_type>
+void ti_upsample_linearNd_kernel_impl(
+    Tensor& output,
+    const Tensor& input,
+    bool align_corners,
+    const scale_type& scales) {
+
+  // input can be NCHW, NCL or NCKHW
+  auto shape = input.sizes().vec();
+  auto strides = input.strides().vec();
+  auto oshape = output.sizes();
+
+  for (int i=0; i<out_ndims; i++) {
+    shape[i + 2] = oshape[i + 2];
+    strides[i + 2] = 0;
+  }
+  auto restrided_input = input.as_strided(shape, strides);
```

Comment:
Can we maybe refer to the advanced indexing implementation to describe what's going on here? This can be tricky to understand for someone new to this implementation.